README.rst
c64-salesforce-client and c64-serial-proxy
==========================

For the Dreamforce 2017 session *New Life for a Commodore 64: An Old-Is-New-Again Salesforce Client*.

https://success.salesforce.com/Sessions#/session/a2q3A000001ytadQAA


See the README in c64-serial-proxy for instructions about the python proxy.


Do make this work as-is, you'll want to install the VICE emulator, and the ACME 6502 assembler.


Build Demo 1 BASIC Client
-------------------------

```
petcat -w2 -o sfdc.prg -- sfdc.bas
```

```
c1541 -format sfdc,id d64 sfdc.d64 -write sfdc.prg sfdc
```


Build Demo 2 Assembly Client
----------------------------

```
acme --cpu 6510 --format cbm --outfile blink.prg macros.asm irpt.asm rs232.asm defs.asm
```

```
c1541 -format blink,id d64 blink.d64 -write blink.prg blink
```


c64-serial-proxy/NEWS.txt
News
====

0.1
---

*Release date: 06-Nov-2017*

* Dreamforce 2017



irpt.asm
;Based on http://dustlayer.com/c64-coding-tutorials/2013/4/8/episode-2-3-did-i-interrupt-you 

BORDERAD = $d020
; We want to flash that green money color
FLASHCOL = 5
UNFLCOL = 14

; number of jiffies we want to wait
FLASHDEL = 40
; number of jiffies/256 we want to wait
SOUNDDEL = 2

; Autorun on load
* = $0801                               ; BASIC start address (#2049)
!byte $0d,$08,$dc,$07,$9e,$20,$34,$39   ; BASIC loader to start at $c000...
!byte $31,$35,$32,$00,$00,$00           ; puts BASIC line 2012 SYS 49152
* = $c000                               ; start address for 6502 code


           ; Open the serial connection
           jsr rs232_open

           ; Clear a timer storage space for the music
           lda #0
           sta $cffd

           ; Set up the irq override
           sei         ; set interrupt disable flag
            
           ldy #$7f    ; $7f = %01111111
           sty $dc0d   ; Turn off CIAs Timer interrupts
           sty $dd0d   ; Turn off CIAs Timer interrupts
           lda $dc0d   ; cancel all CIA-IRQs in queue/unprocessed
           lda $dd0d   ; cancel all CIA-IRQs in queue/unprocessed
          
           lda #$01    ; Set Interrupt Request Mask...
           sta $d01a   ; ...we want IRQ by Rasterbeam

           lda #<irq   ; point IRQ Vector to our custom irq routine
           ldx #>irq 
           sta $314    ; store in $314/$315
           stx $315   

           lda #$00    ; trigger first interrupt at row zero
           sta $d012

           lda $d011   ; Bit#0 of $d011 is basically...
           and #$7f    ; ...the 9th Bit for $d012
           sta $d011   ; we need to make sure it is set to zero 

           cli         ; clear interrupt disable flag
           ; Done setting up the irq override
           jmp *       ; infinite loop


irq        dec $d019        ; acknowledge IRQ

           ; see if we are already in a flash
           lda BORDERAD
           and #15
           cmp #FLASHCOL
           beq chkstopit ; if so, should we stop?
           bne chkserial ; if not, should we start?

done       jsr $f69b ; increment clock
           jsr $579a ; play music for this irq run
           jmp $ea81 ; return to kernel interrupt routine

chkserial  jsr rs232_try_read_byte ; look for byte in serial port
           cmp #0
           bne bkgdon ; we should start
           jmp done

bkgdon     lda $a2 ; bring in system timer value
           sed ; do decimal math
           adc #FLASHDEL ; add the configured delay 
           sta $cfff ; store the time at which we want to stop the flash

           ; change the color
           lda #FLASHCOL
           sta BORDERAD

           ; should we start the sound?
           lda $a1 ; bring in a different byte for the sound timer
           cmp $cffd ; have we ever started the sound, or is it time to start it again?
           bcs soundon ; if $a1 >= $cffd, start the sound
           jmp done

soundon    jsr $5597 ; run the init code for the sid sound file
           lda $a1 ; set up system timer value for sound
           sed ; do decimal math
           adc #SOUNDDEL ; add the configured delay
           sta $cffd ; store the time at which it's ok to restart the music
           jmp done

bkgdoff    lda #0
           sta $cfff ; Clear the byte that holds the flash end time

           ; change the color
           lda #UNFLCOL
           sta BORDERAD
           jmp done

chkstopit  lda $a2
           cmp $cfff ; is the system time past the flash stop time?
           bcs bkgdoff ; if $a2 >= $cfff
           jmp done

* = $558c
!bin "resources/music.sid",,$7c+2


sfdc.bas
   10 open 2,2,0,chr$(10)+chr$(0)
   11 print "loading..."
   12 gosub 59050:rem make "cloud" icon
   13 dim rc$(50)
   14 gosub 16
   15 goto 20
   16 print chr$(147) + "c= client"
   17 return
   20 rem command loop
   21 ep$=""
   22 poke 198,3:poke 631,34:poke 632,34:poke 633,20
   23 input cm$
   24 poke 631,34:poke 632,34:poke 633,20
   25 goto 60000
   100 rem response loop
   111 ur$=""
   120 if st<>0 then print "status " + str$(st)
   122 get#2,c$:if c$="" goto 122
   130 if st=8 goto 120
   131 if asc(c$)=13 goto 150:rem start of corrupt char fixes
   132 if asc(c$)>=192 and asc(c$)<=223 then c$=chr$(asc(c$)-96)
   133 if asc(c$)>=97 and asc(c$)<=122 then c$=chr$(asc(c$)-32)
   134 if asc(c$)>=176 and asc(c$)<=185 then c$=chr$(asc(c$)-128)
   135 if asc(c$)=168 then c$="t"
   136 if asc(c$)=172 then c$=","
   137 if asc(c$)=162 then c$="q"
   138 if asc(c$)=185 then c$="a"
   139 if asc(c$)=182 then c$="v"
   140 if asc(c$)=178 then c$="r"
   141 if asc(c$)=174 then c$="n"
   142 ur$=ur$+c$
   143 goto 120
   150 goto 61000
   1000 rem print record
   1001 a$=fl$:ma=10:sl$=",":cn=1
   1002 gosub 59012:rem split record values for printing
   1003 for k=1 to wd
   1004 print str$(k) + ": " + mid$(vl$, cn, val(wd$(k))) + " ";
   1005 cn=cn+val(wd$(k)):next:print
   1006 return
   59010 rem string splitter, splits twice by default
   59011 ma=3:sl$=" ":rem splitter default settings
   59012 for i=0 to ma: wd$(i)="": next: wd=1
   59013 for i=1 to len(a$)
   59015 wd$(0)=mid$(a$,i,1)
   59017 if wd$(0)=sl$ and wd<>ma then wd=wd+1: next
   59019 wd$(wd)=wd$(wd)+wd$(0): next
   59022 return
   59050 rem replace question mark as input prompt
   59052 poke 52, 48: poke 56, 48
   59053 ug=12288:cg=53248
   59054 poke 56334, peek(56334) and 254
   59055 poke 1, peek(1) and 251
   59056 for k=1 to 2047
   59057 poke ug+k, peek(cg+k)
   59058 next
   59059 poke 1, peek(1) or 4
   59060 poke 56334, peek(56334) or 1
   59061 poke 53272, (peek(53272) and 240)+12
   59062 for j=0 to 7:read a: poke ug+8*63+j, a:next j
   59064 data 000, 000, 056, 124, 126, 254, 048, 000
   59065 return
   60000 rem command dispatch
   60001 rem print cm$:rem echo command
   60002 a$=cm$
   60003 gosub 59010:rem split command
   60004 if wd$(1)="quit" goto 60100
   60005 if ep$<>"" and mid$(wd$(1),1,len(ep$))<>ep$ goto 20
   60006 if wd$(1)="query" goto 60105
   60007 if wd$(1)="update" goto 60120
   60008 if wd$(1)="search" goto 60130
   60009 goto 20
   60100 rem quit command
   60101 end
   60105 rem query command
   60106 print "doing a query"
   60107 print#2,"query:" + wd$(2) + " " + wd$(3) + chr$(0);
   60108 ep$="query"
   60109 goto 100
   60120 rem update command
   60121 print "doing an update"
   60122 print#2,"update:" + wd$(2) + " " + wd$(3) + chr$(0);
   60123 ep$="update"
   60124 goto 100
   60130 rem search command
   60131 print "doing a search"
   60132 print#2,"search:" + wd$(2) + " " + wd$(3) + chr$(0);
   60133 ep$="search"
   60134 goto 100
   61000 rem response dispatch
   61001 rem print "response " + ur$:rem echo response
   61002 a$=ur$
   61003 gosub 59010:rem split command
   61004 if wd$(1)="quit" goto 60100
   61005 if ep$="" then return:rem if we are not expecting a response, ignore
   61006 if wd$(1)="query" or wd$(1)="search" goto 61105
   61007 if wd$(1)="update" goto 61305
   61008 goto 100
   61105 rem query response
   61107 if wd$(2)="count" goto 61220
   61108 if wd$(2)="fields" goto 61240
   61109 if wd$(2)="record" goto 61230
   61110 if wd$(2)="error" goto 61250
   61111 if wd$(2)="done" goto 20
   61120 goto 100
   61220 rem get ready for query/search response
   61221 ri=1:gosub 16
   61223 goto 61120
   61230 rem query/search record handler
   61231 print chr$(18) + chr$(ri+64) + chr$(146) + ": ";
   61232 rc$(ri)=wd$(3):ri=ri+1
   61233 a$=wd$(3):ma=2:sl$=" "
   61234 gosub 59012:rem split record parts for printing
   61235 fl$=wd$(1):vl$=wd$(2)
   61236 gosub 1000
   61239 goto 61120
   61240 rem query fields handler
   61241 a$=wd$(3):ma=10:sl$=" "
   61242 gosub 59012:rem split object name and field names
   61243 print wd$(1)
   61244 for j=2 to wd: print str$(j-1) + ": " + wd$(j) spc(2);:next:print
   61249 goto 61120
   61250 print "query error: " + wd$(3)
   61259 goto 20
   61305 rem update response
   61307 if wd$(2)="error" goto 61350
   61308 if wd$(2)="done" goto 20
   61350 print "update error: " + wd$(3)
   61359 goto 20


rs232.asm
!zone rs232

;.file_name !byte 6, 0
.output_buffer !fill 256, 0
.input_buffer !fill 256, 0
;.tmp !byte 0

; ----------------------------------------------------------------------
; Opens rs232 channel on file #3
; ----------------------------------------------------------------------
rs232_open
		+set16im .input_buffer, RS232_INBUF_PTR
		+set16im .output_buffer, RS232_OUTBUF_PTR

		lda #3          ; file #
		ldx #2          ; 2 = rs-232 device
		ldy #0        ; no cmd
		jsr SETLFS		

		lda #0		; no name
		jsr SETNAM

		lda #%00001000  ; 2400 baud, 8 bits per char
		sta $0293

		jsr OPEN
		rts


; ----------------------------------------------------------------------
; Returns: A
; If no data available, will return immediately with \0 and bit #3 in RSSTAT will be 1
; ----------------------------------------------------------------------
rs232_try_read_byte
		ldx #3
		jsr CHKIN       ; select file 3 as input channel
		jsr GETIN       ; try and read from rs232 buffer
		tay             ; CLRCHN uses A, so move data to Y reg
		jsr CLRCHN
		tya             ; ... and back again
		rts


logokoala.asm
;From http://www.lemon64.com/forum/viewtopic.php?t=36491&sid=0c9db642484183b7b07bf8d768f056c8

;Koala Picture Code by Stranger on CSDb 
PICTURE = $2000 
BITMAP = PICTURE 
VIDEO = PICTURE+$1f40 
COLOR = PICTURE+$2328 
BACKGROUND = PICTURE+$2710 

*=$0801 
; BASIC stub: "1 SYS 2061" 
!by $0b,$08,$01,$00,$9e,$32,$30,$36,$31,$00,$00,$00 

JMP $5000 
* = $5000 
sei 
lda #$00 
sta $d020 ; Border Color 
lda BACKGROUND 
sta $d021 ; Screen Color 

; Transfer Video and Color 
ldx #$00 
.LOOP 
; Transfers video data 
lda VIDEO,x 
sta $0400,x 
lda VIDEO+$100,x 
sta $0500,x 
lda VIDEO+$200,x 
sta $0600,x 
lda VIDEO+$2e8,x 
sta $06e8,x 
; Transfers color data 
lda COLOR,x 
sta $d800,x 
lda COLOR+$100,x 
sta $d900,x 
lda COLOR+$200,x 
sta $da00,x 
lda COLOR+$2e8,x 
sta $dae8,x 
inx 
bne .LOOP 
; 
; Bitmap Mode On 
; 
lda #$3b 
sta $d011 
; 
; MultiColor On 
; 
lda #$d8 
sta $d016 



;Wait Code by PeteD on Lemon64 
lda #$18 
sta $d018 
         ldx #60*3  ;60 fps * 3 seconds 
         lda #128 
loop1 cmp $d012     ;check if the raster has reached line 128 
         bne loop1  ;no, so keep checking 
loop2 cmp $d012     ;if it has you want to make sure you dont catch it more than once per frame 
         beq loop2  ;so wait till it isn't 0 any more 
         dex 
         bne loop1  ;loop round 60*3 times 
          
;File must be at same directory ofcourse 

* = $1ffe 
!binary "resources/qtlfu_ko_fs.koala",,0 


c64-serial-proxy/src/c64serialproxy/__init__.py
from gevent import monkey
monkey.patch_all()

import gevent
import gevent.queue
import click
from serial import serial_for_url
import re
import sys
import fcntl
import os
import socket
from getpass import getpass
from salesforce_requests_oauthlib import SalesforceOAuth2Session
from urllib import urlencode
from salesforce_streaming_client import SalesforceStreamingClient


username = os.environ.get('SALESFORCE_USERNAME')
password = os.environ.get('SALESFORCE_PASSWORD') or getpass(
    'Enter password for {0}: '.format(username)
)

# Yep, a global.
session = None


def search_handler(processor_output_queue, processor_id,
                   previous_command_state, command_args):
    response = session.get(
        '/services/data/vXX.X/parameterizedSearch/',
        params={
            'q': command_args[0],
            'overallLimit': '10',
            'sobject': 'Contact',
            'Contact.fields': 'Name,Phone'
        }
    ).json()

    if 'searchRecords' in response:
        return _query_search_response_builder(
            processor_output_queue,
            processor_id,
            'search',
            len(response['searchRecords']),
            response['searchRecords']
        )
    else:
        # Sorry for not duck typing
        payload = {
            'event': 'search error',
            'processor_id': processor_id
        }
        if isinstance(response, list):
            payload['error'] = response[0][u'message']
        else:
            print response
        processor_output_queue.put(payload)


def update_handler(processor_output_queue, processor_id,
                   previous_command_state, command_args):
    record_handle, field_handle, value = command_args[0].split(None, 2)
    field_handle = int(field_handle)
    response = session.patch(
        '/services/data/vXX.X/sobjects/{0}/{1}'.format(
            previous_command_state['sobject_name'],
            previous_command_state['Ids'][record_handle]
        ),
        json={
            previous_command_state['fields'][field_handle]: value
        }
    )
    if response.status_code == 204:
        processor_output_queue.put({
            'event': 'update done',
            'processor_id': processor_id
        })
    else:
        processor_output_queue.put({
            'event': 'update error',
            'processor_id': processor_id
        })


def query_handler(processor_output_queue, processor_id, previous_command_state,
                  command_args):
    response = session.get('/services/data/vXX.X/query/?{0}'.format(
        urlencode({'q': command_args[0]})
    )).json()
    # Ignoring done for now
    if 'records' in response:
        return _query_search_response_builder(
            processor_output_queue,
            processor_id,
            'query',
            response['totalSize'],
            response['records']
        )
    else:
        # Sorry for not duck typing
        payload = {
            'event': 'query error',
            'processor_id': processor_id
        }
        if isinstance(response, list):
            payload['error'] = response[0][u'message']
        else:
            print response
        processor_output_queue.put(payload)


def _query_search_response_builder(processor_output_queue, processor_id,
                                   query_or_search, size, records):
        processor_output_queue.put({
            'event': '{0} count'.format(query_or_search),
            'processor_id': processor_id,
            'count': size
        })

        to_return = {
            'Ids': {},
            'fields': {}
        }
        for i in xrange(len(records)):
            record = records[i]
            sobject_type = record[u'attributes'][u'type']
            to_return['Ids'][chr(97 + i)] = \
                record[u'attributes'][u'url'].rsplit('/', 1)[1]
            del record[u'attributes']

            if i == 0:
                to_return['sobject_name'] = sobject_type

                sorted_keys = sorted(record.keys())
                to_return['fields'] = dict(enumerate(sorted_keys, 1))

                # Gather field names so we can send them along
                processor_output_queue.put({
                    'event': '{0} fields'.format(query_or_search),
                    'processor_id': processor_id,
                    'fields': [name.encode('ascii', errors='replace')
                               for name
                               in sorted_keys],
                    'type': sobject_type
                })

            processor_output_queue.put({
                'event': '{0} record'.format(query_or_search),
                'processor_id': processor_id,
                'record': record
            })
        processor_output_queue.put({
            'event': '{0} done'.format(query_or_search),
            'processor_id': processor_id
        })

        return to_return


# Handlers should put responses suitable for the c64 to handle into
# processor_output_queue
command_map = {
    'query': query_handler,
    'search': search_handler,
    'update': update_handler
}

re_allowed_chars = re.compile(
    r'[^A-Za-z\xc1-\xda\x11-\x14 0-9.,?\'"!@#$%^&*()-_=+;:<>/\\|}{[\]`~\r]*'
)


def to_petscii(inp):
    inp = inp.replace(r'\r', '')
    inp = inp.replace(r'\n', '\r')
    inp = re_allowed_chars.sub('', inp)
    inp = inp.replace(r'_', '-')
    inp = inp.replace(r'`', '\x27')  # '`' in petscii

    petscii_inp = []
    for char in inp:
        char_ord = ord(char)
        if 65 <= char_ord <= 90:
            petscii_inp.append(chr(char_ord + 32))
        elif 97 <= char_ord <= 122:
            petscii_inp.append(chr(char_ord - 32))
        elif 193 <= char_ord <= 218:
            petscii_inp.append(chr(char_ord - 128))
        else:
            petscii_inp.append(char)

    return ''.join(petscii_inp)


def from_petscii(inp):
    return to_petscii(inp.replace(r'\r', '\n')).replace(r'\r', '\n')


def write(control, serial_conn, write_queue):
    while not control['die']:
        to_write = None
        with gevent.Timeout(2, False) as timeout:
            to_write = write_queue.get()
        if to_write is not None:
            to_write = to_petscii(to_write)
            print 'writing', to_write
            for char in to_write:
                serial_conn.write(char)
                gevent.sleep(0.1)
            gevent.sleep(0.5)
            serial_conn.flush()


def read(control, serial_conn, command_queue):
    command = []
    while not control['die']:
        char = None
        with gevent.Timeout(2, False) as timeout:
            char = serial_conn.read()
        if char is None:
            continue
        if char == '\x00':
            command_queue.put(''.join(command))
            command = []
        else:
            command.append(from_petscii(char))


def gather_processor_output(control, processor_output_queue, write_queue):
    while not processor_output_queue.empty() or not control['die']:
        processor_output = None
        with gevent.Timeout(2, False) as timeout:
            processor_output = processor_output_queue.get()
        if processor_output is not None:
            print 'Processor {0}: {1}'.format(
                processor_output['event'],
                processor_output['processor_id']
            )

            # For sobjects returned by a rest api call
            event_tokens = processor_output['event'].split()
            if event_tokens[0] == 'query' or event_tokens[0] == 'search':
                if event_tokens[1] == 'count':
                    write_queue.put('{0} count {1}\\n'.format(
                        event_tokens[0],
                        str(processor_output['count'])
                    ))
                elif event_tokens[1] == 'fields':
                    write_queue.put('{0} fields {1} {2}\\n'.format(
                        event_tokens[0],
                        processor_output['type'],
                        ' '.join(processor_output['fields'])
                    ))
                elif event_tokens[1] == 'record':
                    write_queue.put('{0} record '.format(
                        event_tokens[0]
                    ))
                    record = processor_output['record']

                    value_lengths = []
                    values = []
                    for field_name in sorted(record.keys()):
                        value = record[field_name]
                        if value is None:
                            value = ''
                        value = value.encode('ascii', errors='replace')

                        value_lengths.append(str(len(value)))
                        values.append(value)
                    write_queue.put(','.join(value_lengths))
                    write_queue.put(' ')
                    write_queue.put(''.join(values))
                    write_queue.put('\\n')
                elif event_tokens[1] == 'error':
                    write_queue.put('{0} error {1}\\n'.format(
                        event_tokens[0],
                        processor_output['error']
                    ))
                else:
                    write_queue.put('{0}\\n'.format(processor_output['event']))
            else:
                write_queue.put('{0}\\n'.format(processor_output['event']))


# Runs in the main greenlet
# We expect these command handlers to spend most of their time
# doing network stuff, so we don't need a child process.
def command_dispatch(control, command_queue, write_queue):
    processor_output_queue = gevent.queue.Queue()
    processor_output_greenlet = gevent.spawn(
        gather_processor_output,
        control,
        processor_output_queue,
        write_queue
    )
    processor_id = 0
    previous_command_state = None
    while not control['die']:
        command = ''
        with gevent.Timeout(5, False) as timeout:
            command = command_queue.get()
        command_segments = command.split(':', 1)
        if command_segments[0] in command_map:
            processor = gevent.spawn(
                command_map[command_segments[0]],
                processor_output_queue,
                processor_id,
                previous_command_state,
                command_segments[1:]
            )
            processor_id += 1

            # Don't loop again until this processor is done.  We could store
            # these and join the list later, but for now we'd rather simplify
            # the storage of state.
            processor.join()
            previous_command_state = processor.value

    # Don't miss any processor output
    processor_output_greenlet.join()


def get_stdin(control, write_queue):
    fcntl.fcntl(sys.stdin, fcntl.F_SETFL, os.O_NONBLOCK)
    while True:
        try:
            gevent.socket.wait_read(sys.stdin.fileno(), 2)
        except socket.timeout:
            continue
        else:
            line = sys.stdin.readline().strip()
            if line.lower() == 'control-quit':
                control['die'] = True
                break
            write_queue.put('{0}\\n'.format(line))


class C64StreamingClient(SalesforceStreamingClient):
    def __init__(self, client_id, client_secret, username,
                 control, write_queue, sandbox=False,
                 replay_client_id=None):

        self.control = control
        self.write_queue = write_queue

        super(C64StreamingClient, self).__init__(
            client_id,
            client_secret,
            username,
            sandbox=sandbox,
            replay_client_id=replay_client_id
        )

    def send_serial(self, connect_response_element):
        self.write_queue.put('l\\n')


@click.command()
@click.argument('device_name')
@click.option('-s', '--streaming', is_flag=True)
def main(device_name, streaming):
    oauth_client_id = os.environ.get('SALESFORCE_CLIENT_ID')
    oauth_client_secret = os.environ.get('SALESFORCE_CLIENT_SECRET')

    write_queue = gevent.queue.Queue()
    command_queue = gevent.queue.Queue()

    control = {
        'die': False
    }

    serial_conn = serial_for_url(device_name, baudrate=1200)
    serial_conn.write('\x00\\n')

    reader_greenlet = gevent.spawn(read, control, serial_conn, command_queue)
    writer_greenlet = gevent.spawn(write, control, serial_conn, write_queue)
    stdin_greenlet = gevent.spawn(get_stdin, control, write_queue)

    all_greenlets = [reader_greenlet, writer_greenlet, stdin_greenlet]
    try:
        if streaming:
            with C64StreamingClient(
                oauth_client_id, oauth_client_secret, username, control,
                write_queue
            ) as streaming_client:
                streaming_client.subscribe(
                    '/event/SoldSomething__e',
                    'send_serial'
                )
                streaming_client.start()
                streaming_client.block()
        else:
            global session
            session = SalesforceOAuth2Session(
                oauth_client_id,
                oauth_client_secret,
                username,
                sandbox=False,
                password=password
            )
            command_dispatch(control, command_queue, write_queue)
    except KeyboardInterrupt:
        pass
    try:
        print 'Stopping...'
        gevent.joinall(all_greenlets)
    except KeyboardInterrupt:
        print [bool(glet) for glet in all_greenlets]
